### PR TITLE
feat(#1134): adds the Epoxy extension to the FlyingSquid label model. 

### LINKED ISSUE #1134: [Weak supervision] Extend FlyingSquid with Epoxy
[Epoxy](https://github.com/HazyResearch/epoxy) should be integrated into the existing FlyingSquid implementation / documentation.

### PR DESCRIPTION
**Pull request by ruanchaves**
Closes #1134.   

Extends weak labels with Epoxy.

Roadmap:

- [x] Extend Weak Labels objects with [Epoxy](https://github.com/HazyResearch/epoxy). Make it work independently from FlyingSquid.
- [x] Perform experiments with distinct label models ( FlyingSquid, Snorkel ), word embeddings ( GPT-3 , sentence-transformers ) and text classification datasets ( ag_news, SST-2, YouTube Spam dataset ).
- [x] Write a tutorial with the results.

### CODE REVIEW

**Comment by dcfidalgo** [comment]
For the discussion today:
- remove `WeakLabelsEmbeddings`
- `Epoxy.__init__(weak_labels, embeddings, thresholds)`
- make classmethods normal methods
- ideal flow would be:
```python
embeddings = []
for rec in weak_labels.records:
    embeddings.append(...)
epoxy = Epoxy(weak_labels, embeddings)
epoxy.fit() #  automatically runs the grid search to find the thresholds
# or
epoxy = Epoxy(weak_labels, embeddings, thresholds)
epoxy.fit() #  omits the grid search
```
- save the best model of the grid search instead of fitting again
- think about what to optimize for in the grid search

**Comment by ruanchaves** [comment]
We're changing the focus of this issue to minimize code changes to the core library and to make Epoxy work independently from FlyingSquid. I have updated the roadmap accordingly.

File: src/rubrix/labeling/text_classification/weak_labels.py [hunk 1]
```diff
@@ -403,12 +407,110 @@ def matrix(self, has_annotation: Optional[bool] = None) -> np.ndarray:
...
+        ]
+        nearest = [
+            dist_and_nearest[1].flatten() for dist_and_nearest in dists_and_nearest
+        ]
+
+        return dists, nearest
+
+    def extend_matrix(
+        self, thresholds, embeddings: Optional[np.ndarray] = None, gpu=False
+    ) -> np.ndarray:
```
Comment (dcfidalgo) [inline src/rubrix/labeling/text_classification/weak_labels.py]:
```suggestion
    ):
```

File: src/rubrix/labeling/text_classification/weak_labels.py [hunk 2]
```diff
@@ -403,12 +407,110 @@ def matrix(self, has_annotation: Optional[bool] = None) -> np.ndarray:
...
+        Args:
+            embeddings: sentence embeddings for each row of the weak labels matrix.
+            thresholds: thresholds for the minimum cosine similarity between sentences for a rule to be expanded.
+            gpu: Perform FAISS similarity queries on GPU.
+            has_annotation: If True, return only the part of the matrix that has a corresponding annotation.
+                If False, return only the part of the matrix that has NOT a corresponding annotation.
+                By default, we return the whole weak label matrix.
+
+        Returns:
+            The extended weak label matrix, or optionally just a part of it.
```
Comment (dcfidalgo) [inline src/rubrix/labeling/text_classification/weak_labels.py]:
```suggestion
```

File: src/rubrix/labeling/text_classification/weak_labels.py [hunk 3]
```diff
@@ -403,12 +407,110 @@ def matrix(self, has_annotation: Optional[bool] = None) -> np.ndarray:
...
+        support = []
+        for i in range(matrix_length):
+            support.append(np.argwhere(self._matrix[:, i] != none_label_int).flatten())
+
+        mat_abstains = [
+            np.argwhere(self._matrix[:, i] == none_label_int).flatten()
+            for i in range(matrix_length)
+        ]
+
+        if not embeddings is None:
```
Comment (dcfidalgo) [inline src/rubrix/labeling/text_classification/weak_labels.py]:
```suggestion
        if embeddings is not None:
```

File: src/rubrix/labeling/text_classification/weak_labels.py [hunk 4]
```diff
@@ -403,12 +407,110 @@ def matrix(self, has_annotation: Optional[bool] = None) -> np.ndarray:
...
+        nearest = [
+            dist_and_nearest[1].flatten() for dist_and_nearest in dists_and_nearest
+        ]
+
+        return dists, nearest
+
+    def extend_matrix(
+        self, thresholds, embeddings: Optional[np.ndarray] = None, gpu=False
+    ) -> np.ndarray:
+        """Returns the extended weak label matrix, or optionally just a part of it.
```
Comment (dcfidalgo) [inline src/rubrix/labeling/text_classification/weak_labels.py:464]: Revise doc string

**Review by dcfidalgo — COMMENTED** [review]
Just a few small things i saw during my quick review

**Comment by dcfidalgo** [comment]
Closing this one in favor of #1487, rebasing this was a nightmare.

**Comment by ruanchaves** [comment]
Closing this long PR due to merge conflicts. Future commits will be made at #1487 .

**Pull request closed without merging**